lib/common/asar.js

```diff
@@ -39,202 +35,132 @@
   // Separate asar package's path from full path.
   const splitPath = function (p) {
     // shortcut to disable asar.
-    if (isAsarDisabled()) {
-      return [false]
-    }
-
-    if (Buffer.isBuffer(p)) {
-      p = p.toString()
-    }
-
-    if (typeof p !== 'string') {
-      return [false]
-    }
-
-    if (p.substr(-5) === '.asar') {
-      return [true, p, '']
-    }
+    if (isAsarDisabled()) return [false]
+    if (Buffer.isBuffer(p)) p = p.toString()
+    if (typeof p !== 'string') return [false]
+    if (p.substr(-5) === '.asar') return [true, p, '']
 
     p = path.normalize(p)
+
     const index = p.lastIndexOf('.asar' + path.sep)
-    if (index === -1) {
-      return [false]
-    }
+    if (index === -1) return [false]
+
     return [true, p.substr(0, index + 5), p.substr(index + 6)]
   }
 
   // Convert asar archive's Stats object to fs's Stats object.
   let nextInode = 0
 
   const uid = process.getuid != null ? process.getuid() : 0
-
   const gid = process.getgid != null ? process.getgid() : 0
 
   const fakeTime = new Date()
+  const msec = (date) => (date || fakeTime).getTime()
 
   const asarStatsToFsStats = function (stats) {
-    return {
-      dev: 1,
-      ino: ++nextInode,
-      mode: 33188,
-      nlink: 1,
-      uid: uid,
-      gid: gid,
-      rdev: 0,
-      atime: stats.atime || fakeTime,
-      birthtime: stats.birthtime || fakeTime,
-      mtime: stats.mtime || fakeTime,
-      ctime: stats.ctime || fakeTime,
-      size: stats.size,
-      isFile: function () {
-        return stats.isFile
-      },
-      isDirectory: function () {
-        return stats.isDirectory
-      },
-      isSymbolicLink: function () {
-        return stats.isLink
-      },
-      isBlockDevice: function () {
-        return false
-      },
-      isCharacterDevice: function () {
-        return false
-      },
-      isFIFO: function () {
-        return false
-      },
-      isSocket: function () {
-        return false
-      }
-    }
+    const { Stats, constants } = require('fs')
+
+    let mode = constants.S_IROTH ^ constants.S_IRGRP ^ constants.S_IRUSR ^ constants.S_IWUSR
+
+    if (stats.isFile) {
+      mode ^= constants.S_IFREG
+    } else if (stats.isDirectory) {
+      mode ^= constants.S_IFDIR
+    } else if (stats.isLink) {
+      mode ^= constants.S_IFLNK
+    }
```

Comment:
We implemented this functionality in a different way before (overriding `isFile` and `isDirectory` and friends).  This PR adjusts the `mode` so that the node.js built in methods can do that work for us.  We already have tests for those 👍